syntax = "proto3";

package ecliptix.proto.protocol;
option csharp_namespace = "Ecliptix.Protobuf.Protocol";

import "google/protobuf/timestamp.proto";

message ProtocolState {
  uint32 version = 1; // must be 1
  bool is_initiator = 2;
  google.protobuf.Timestamp created_at = 3;
  bytes session_id = 4; // 16 bytes
  uint64 state_counter = 5; // monotonic, app should store to detect rollback
  uint64 send_ratchet_epoch = 6;
  uint64 recv_ratchet_epoch = 7;

  bytes root_key = 10; // 32 bytes
  ChainState send_chain = 11;
  ChainState recv_chain = 12;

  DhKeyPair dh_local = 13; // current sending DH key pair
  bytes dh_remote_public = 14; // current peer DH public key (32 bytes)
  bytes dh_local_initial_public = 15; // immutable (32 bytes)
  bytes dh_remote_initial_public = 16; // immutable (32 bytes)

  bytes metadata_key = 17; // 32 bytes, stable for the session

  KyberKeyPair kyber_local = 18;
  bytes kyber_remote_public = 19; // 1184 bytes

  NonceState nonce_generator = 20;
  bytes state_hmac = 21; // 32 bytes, HMAC-SHA256 over deterministic state serialization with state_hmac empty
  uint32 max_messages_per_chain = 22; // per-session chain length
}

message ChainState {
  uint64 message_index = 1; // next message index
  bytes chain_key = 2; // 32 bytes
  repeated CachedMessageKey skipped_message_keys = 3; // optional
}

message CachedMessageKey {
  uint64 message_index = 1;
  bytes message_key = 2; // 32 bytes
}

message DhKeyPair {
  bytes private_key = 1; // 32 bytes
  bytes public_key = 2;  // 32 bytes
}

message KyberKeyPair {
  bytes secret_key = 1; // 2400 bytes
  bytes public_key = 2; // 1184 bytes
}

message NonceState {
  bytes prefix = 1; // 4 bytes
  uint64 counter = 2;
}
